### PR TITLE
Add authority.WithX509SignerFunc

### LINKED ISSUE #874: Add authority option WithX509SignerFunc
### Description

Caddy proposes adding a new function to be able to automatically replace the intermediate certificate in an authority using a function that returns the current certificate chain and signer to use. See https://github.com/caddyserver/caddy/issues/4517.

The new option can be something like

```golang
func WithX509SignerFunc(func() (chain []*x509.Certificate, s crypto.Signer)) Option
or 
func WithX509SignerFunc(func() (chain []*x509.Certificate, s crypto.Signer, err error)) Option

```

Then `SoftCAS` if configured can use that to get the updated chain and signer.

### PR DESCRIPTION
**Pull request by maraino**
### Description

This change adds a new authority option that allows passing a callback that returns the certificate chain and signer used to sign X.509 certificates.

This option will be used by Caddy, they renew the intermediate certificate weekly and there's no other way to replace it without re-creating the embedded CA.

Fixes #874

### CODE REVIEW

**Review by azazeal — APPROVED** [review]
LGTM - minor comments

File: cas/softcas/softcas_test.go [hunk 1]
```diff
@@ -186,6 +192,10 @@ func setTeeReader(t *testing.T, w *bytes.Buffer) {
 }
 
 func TestNew(t *testing.T) {
+	assertEqual := func(x, y interface{}) bool {
```
[the remark this answers is not in the document]
Comment (maraino) [reply cas/softcas/softcas_test.go:195]: You cannot compare functions, even if they are the same. I don't know if testify has something hacky like that. You can do it with go-cmp.Equal ignoring a specific type or parameter, but that parameter is the only one I want to check.

[the remark this answers is not in the document]
Comment (azazeal) [reply cas/softcas/softcas_test.go:195]: didn't look that far ahead - just saw the assertion!

File: cas/apiv1/options.go [hunk 2]
```diff
@@ -31,13 +31,18 @@ type Options struct {
 	// https://cloud.google.com/docs/authentication.
 	CredentialsFile string `json:"credentialsFile,omitempty"`
 
-	// Certificate and signer are the issuer certificate, along with any other
-	// bundled certificates to be returned in the chain for consumers, and
+	// CertificateChain and Signer are the issuer certificate, along with any
```
[the remark this answers is not in the document]
Comment (maraino) [reply cas/apiv1/options.go]: Fixed with c480936

**Comment by hussamabid** [comment]
[]()

**Pull request merged**